src/snosearch/queries.py

```diff
@@ -160,12 +160,27 @@ def _get_subtypes_for_item_type(self, item_type):
     def _get_name_for_item_type(self, item_type):
         return self._get_registered_types()[item_type].name
 
-    def _get_collection_name_for_item_type(self, item_type):
-        return getattr(
-            self._get_registered_types()[item_type],
-            COLLECTION_NAME,
-            None
-        )
+    def _get_collection_names_for_item_type(self, item_type):
```

Comment:
I think this can be cleaned up a lot. 

Instead of touching this function can you add a new method, something like `_resolve_abstract_types_to_subtypes(self, item_types)`. Then just call the `_get_subtypes_for_item_type(item_type)` method on all the item types and flatten the list? https://github.com/IGVF-DACC/snovault-search/blob/d9d3bb7bad4cc84d37d79a1965ae06518530ac19/src/snosearch/queries.py#L157-L158

We want to do this just for indices, so really the only thing that should change is wrapping this https://github.com/IGVF-DACC/snovault-search/blob/d9d3bb7bad4cc84d37d79a1965ae06518530ac19/src/snosearch/queries.py#L122-L124 with the `_resolve_abstract_types_to_subtypes` method. Then the normal `_get_collection_names_for_item_types` should work without modification.

So it should look something like this:

```
def _resolve_abstract_types_to_subtypes(self, item_types):
    subtypes = set()
    for item_type in item_types:
        subtypes.update(self._get_subtypes_for_item_type(item_type))
    return list(sorted(subtypes))
```

The `_get_subtypes_for_item_type` method always returns a list of concrete types. If you call it on a concrete type it just returns itself. So basically we take the type parameters passed in and run them through this resolver function. If the type is abstract it gets resolved into its concrete subtypes, if it's concrete it just gets passed through. We are deduplicating with a set and sorting into a list just to keep things consistent but shouldn't really matter.

Then in the `_get_index` method we just wrap the passed in type parameters with the resolver function:

```
 def _get_index(self):
        if self._should_search_over_all_indices():
            return [RESOURCES_INDEX]
        return self._get_collection_names_for_item_types(
            self._resolve_abstract_types_to_subtypes( # <------ wrapper here
                self.params_parser.param_values_to_list(
                    params=self._get_item_types()
                )
            )
            or self.params_parser.param_values_to_list(
                params=self._get_default_item_types()
            )
        )
```

We shouldn't wrap the default item types, because those are defined by us and should already be concrete. And we shouldn't have to worry about type validation because this already happens as first step: https://github.com/IGVF-DACC/snovault-search/blob/d9d3bb7bad4cc84d37d79a1965ae06518530ac19/src/snosearch/queries.py#L1058

Reply:
Sounds good. I will work on it. The only thing is that the abstract type can have another abstract type as its subtype, that is why I am using recursive method to get all the concrete types for the abstract type. For example, sample is a abstract type, and it has a biosample subtype, which is also a abstract type.

Reply:
It looks like the way subtypes are calculated only deal with concrete objects (https://github.com/IGVF-DACC/snovault/blob/b87c8566f0900bc90c8987e00b746bfe406e891f/src/snovault/typeinfo.py#L37-L41) so you shouldn't have to worry about recursively descending from the abstract parents. It should give you all of the concrete children in one call.

You should assume that the item type is already valid by the time it actually reaches the internal methods. For example it's blocked in real query by separate validation step: https://www.encodeproject.org/search/?type=NotAType

Reply:
Got it. Thanks so much for the help!